### PR TITLE
[RFC] Mbed utee

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This is a proof of concept integration of mbedtls in libutee.

This will make the mbedtls interface available for a TA to use, for instance to verify an X.509 certificate.

This would also be a starting point to implement the GlobalPlatform Annex C: Sockets TLS Spec of TEE Sockets API Specification.

The integration could be optimized to use the GP Internal Core API for different crypto algorithms.

### CODE REVIEW

**Comment by jforissier** [comment]
Why not a separate library (libmbedtls.a)?

**Comment by jenswi-linaro** [comment]
If we use it to implement TLS Sockets interface there will be a cross dependency with libutee, unless we put the TLS Sockets stuff in a separate lib too.

**Comment by jforissier** [comment]
Which cross-dependency? I see how libutee with TLS sockets would depend on libmbetls, but how would libmbedtls depend on libutee? Compare the situation to libmpa for instance which is a separate library.

**Comment by jenswi-linaro** [comment]
I expect that in the end we'll implement most of the algorithms used in mbedtls with the ones provided through GP Internal Core API. For instance some keys may not leave kernel mode.

**Comment by jforissier** [comment]
OK I see. It should not be a problem however because we have `--start-group ... --end-group` in the linker flags so circular dependencies should not be an issue (i.e., "it should just work")?

**Comment by jenswi-linaro** [comment]
Yes, I realize that. But if libutee always will need libmbedtls and vice versa, why not make it easy and have everything in one lib?

**Comment by jforissier** [comment]
> But if libutee always will need libmbedtls and vice versa, why not make it easy and have everything in one lib?

Agreed. So let's step back: what if we move the sockets stuff to its own library? The dependencies would become simpler, and TAs would not have to link against code that they don't use if they don't want to.

**Comment by jenswi-linaro** [comment]
Makes sense. So the dependency graph will be: libtlssocket -> libmbedtls -> libutee
Then there's the question on how to share the code with the kernel mode mbedtls integration, but that's another problem.

**Comment by jenswi-linaro** [comment]
Moved mbedtls into a separate library as discussed.
Note the two new config flags for mbedTLS

**Comment by jenswi-linaro** [comment]
I think we should wait with merging the mbedTLS lib until there's an actual TA that will use it.

**Comment by jenswi-linaro** [comment]
Added a user for this pull request with https://github.com/OP-TEE/optee_test/pull/279
This is now a candidate to merge once the review is completed.

**Comment by jforissier** [comment]
Libmbedtls does not follow the kernel coding conventions so it should be excluded from checkpatch.

Could libmbedtls/ be imported as a submodule?

**Comment by edison-ai** [comment]
I think it can coexist with my PR #1885. We can configure these 2 configure macros to indicate the actual mbedtls source patch:
 `CFG_CRYPTOLIB_NAME=mbedtls` and `CFG_CRYPTOLIB_DIR=lib/libmbedtls`
For that, I need refine my patch to use the `CFG_CRYPTOLIB_DIR` for source file build.

But now, I have some changing patch need add to mbedtls for integration.
1. For context memory operation: https://github.com/OP-TEE/optee_os/pull/1885/commits/e2506f9f5c069c383520c0e82f410c0fb39d03a4
2. For GCM and CCM enhance: https://github.com/OP-TEE/optee_os/pull/1885/commits/05471065f8f57b9de195ccea0d15b7229e5e47f5

So, I prepare copy source files in here directly now.

**Comment by jenswi-linaro** [comment]
> Libmbedtls does not follow the kernel coding conventions so it should be excluded from checkpatch.

Updated

> Could libmbedtls/ be imported as a submodule?

Not in the current state, we will need to make some minor modifications for configurations.

File: core/lib/libtomcrypt/src/mpa_desc.c [hunk 1]
```diff
@@ -604,7 +604,7 @@ static int isprime(void *a, int b, int *c)
 	return CRYPT_OK;
 }
 
-static int rand(void *a, int size)
+static int myrand(void *a, int size)
```
Comment (jbech-linaro) [inline core/lib/libtomcrypt/src/mpa_desc.c]: How about mpa_rand? Makes it a bit unique, but still tied to the MPA library?

Comment (jenswi-linaro) [reply core/lib/libtomcrypt/src/mpa_desc.c]: OK, I'll fix.

**Comment by jenswi-linaro** [comment]
Rebased, squashed and tag applied

**Comment by etienne-lms** [comment]
no s-o-b or acked-by for the mbedTLS import commit is ok?

**Comment by jenswi-linaro** [comment]
@etienne-lms, good catch. I'll update.

**Comment by jenswi-linaro** [comment]
Update with s-o-b on "libutee: import mbedtls-2.6.1"

**Comment by jforissier** [comment]
> no s-o-b or acked-by for the mbedTLS import commit is ok?

No, and I still would like to try importing as a submodule instead. At last Connect (HKG18) we said we'd like to rework the LibTomCrypt integration in that way, so that it is easier to track our local changes and rebase when new releases are made available. The same arguments apply to Mbedtls I think.

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jenswi-linaro** [comment]
Replaced by https://github.com/OP-TEE/optee_os/pull/2335

**Pull request closed without merging**